### PR TITLE
add index metrics

### PR DESCRIPTION
**Pull request by matsumana**
I would like to monitor index metrics, just like Kinaba X-pack monitoring.
so I added 3 metrics below:

```
# HELP elasticsearch_indices_docs_primary Count of documents which only primary shards
# TYPE elasticsearch_indices_docs_primary gauge
elasticsearch_indices_docs_primary{cluster="test",index="foo_1"} 2
elasticsearch_indices_docs_primary{cluster="test",index="foo_2"} 3

# HELP elasticsearch_indices_store_size_bytes_primary Current total size of stored index data in bytes which only primary shards on all nodes
# TYPE elasticsearch_indices_store_size_bytes_primary gauge
elasticsearch_indices_store_size_bytes_primary{cluster="test",index="foo_1"} 8425
elasticsearch_indices_store_size_bytes_primary{cluster="test",index="foo_2"} 12420

# HELP elasticsearch_indices_store_size_bytes_total Current total size of stored index data in bytes which all shards on all nodes
# TYPE elasticsearch_indices_store_size_bytes_total gauge
elasticsearch_indices_store_size_bytes_total{cluster="test",index="foo_1"} 16850
elasticsearch_indices_store_size_bytes_total{cluster="test",index="foo_2"} 24840
```

### CODE REVIEW

**Comment by matsumana** [comment]
Hello, Please review.

**Comment by matsumana** [comment]
Hello, Thank you for the comments.

> the metrics your are using from _all/_stats are all included in _nodes/_local/stats/_nodes/stats which we already get from ES in `collector/nodes.go. 

Certainly we can get the metrics of docs from `_nodes/_local/stats` / `_nodes/stats`.
however, which is the docs of primary + replica. I want to collect metrics of primary docs only.
Actually, it seems that Kibana X-Pack monitoring is so.

**Comment by matsumana** [comment]
So I fixed below:

- moved all indices metric from `collector/nodes.go` to `collector/indices.go`
- gathered the metrics collect logic in metrics_collector. To avoid redundant api calls.

What do you think?

**Comment by zwopir** [comment]
not sure if moving the index metrics to a another call to the ES api is what we want. We need to call `_nodes/_local/stats` or depending on the `all`-flag `_nodes/stats` for the nodes-metrics anyways. I would like to make the `indices`-metrics optional, since people might have many indices and we thus create a high metrics cardinality and much more json to transfer and parse.

What do you think, @dominikschulz ?

**Comment by dominikschulz** [comment]
Sounds reasonable. We should make indices metrics optional.

**Comment by matsumana** [comment]
@zwopir @dominikschulz 
I agree with you.
Actually, I have heard that there are Elasticsearch clusters with several hundreds indexes in some companies.
I fixed this PR. Please review.

File: collector/cluster_health.go [hunk 1]
```diff
@@ -220,49 +217,14 @@ func (c *ClusterHealth) Describe(ch chan<- *prometheus.Desc) {
...
-	if err := json.NewDecoder(res.Body).Decode(&chr); err != nil {
-		c.jsonParseFailures.Inc()
-		return chr, err
-	}
-
-	return chr, nil
-}
-
-func (c *ClusterHealth) Collect(ch chan<- prometheus.Metric) {
+func (c *ClusterHealth) Collect(ch chan<- prometheus.Metric, clusterHealthResponse clusterHealthResponse) {
```
Comment (zwopir) [inline collector/cluster_health.go]: this doesn't work. Collect must fulfill the prometheus. Collector interface and thus the signature must be
```
Collect(ch chan<- prometheus.Metric)
```

**Comment by dominikschulz** [comment]
@matsumana IMHO you're changing/moving a lot of unrelated code that shouldn't be touched by your PR.

File: collector/cluster_health.go [hunk 2]
```diff
@@ -220,49 +217,14 @@ func (c *ClusterHealth) Describe(ch chan<- *prometheus.Desc) {
 	ch <- c.jsonParseFailures.Desc()
 }
 
-func (c *ClusterHealth) fetchAndDecodeClusterHealth() (clusterHealthResponse, error) {
```
Comment (dominikschulz) [inline collector/cluster_health.go]: this method should remain in this file

File: collector/cluster_health_test.go [hunk 3]
```diff
@@ -1,55 +0,0 @@
...
-	"fmt"
-	"net/http"
-	"net/http/httptest"
-	"net/url"
-	"testing"
-
-	"github.com/go-kit/kit/log"
-)
-
-func TestClusterHealth(t *testing.T) {
```
Comment (dominikschulz) [inline collector/cluster_health_test.go]: this method should remain in this file

File: collector/metrics_collector.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	clusterHealth *ClusterHealth
+	nodes         *Nodes
+	indices       *Indices
+
+	clusterHealthResponse *clusterHealthResponse
+	nodeStatsResponse     *nodeStatsResponse
+	indexStatsResponse    *indexStatsResponse
+}
+
+func NewMetricsCollector(logger log.Logger, client *http.Client, url *url.URL, all bool, exportIndices bool) *MetricsCollector {
```
Comment (dominikschulz) [inline collector/metrics_collector.go]: We should init all metrics collector in main, not in another wrapper method.

File: collector/metrics_collector.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+		return
+	}
+	c.indices.up.Set(1)
+
+	c.clusterHealth.Collect(ch, clusterHealthResponse)
+	c.nodes.Collect(ch, nodeStatsResponse)
+	c.indices.Collect(ch, clusterHealthResponse, nodeStatsResponse, indexStatsResponse)
+}
+
+func (c *ClusterHealth) fetchAndDecodeClusterHealth() (clusterHealthResponse, error) {
```
Comment (dominikschulz) [inline collector/metrics_collector.go]: this method should stay in it's own file.

File: collector/metrics_collector.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+	if err := json.NewDecoder(res.Body).Decode(&chr); err != nil {
+		c.jsonParseFailures.Inc()
+		return chr, err
+	}
+
+	return chr, nil
+}
+
+func (c *Nodes) fetchAndDecodeNodeStats() (nodeStatsResponse, error) {
```
Comment (dominikschulz) [inline collector/metrics_collector.go]: this method should stay in it's own file.

File: collector/metrics_collector_test.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+		w.WriteHeader(401)
+		w.Write([]byte("401 Unauthorized\n"))
+		return
+	}
+
+	h.Next.ServeHTTP(w, r)
+	return
+}
+
+func TestClusterHealth(t *testing.T) {
```
Comment (dominikschulz) [inline collector/metrics_collector_test.go]: this method should stay in it's own file.

File: collector/metrics_collector_test.go [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+		if chr.NumberOfNodes != 1 {
+			t.Errorf("Wrong number of nodes")
+		}
+		if chr.NumberOfDataNodes != 1 {
+			t.Errorf("Wrong number of data nodes")
+		}
+	}
+}
+
+func TestNodesStats(t *testing.T) {
```
Comment (dominikschulz) [inline collector/metrics_collector_test.go]: this method should stay in it's own file.

File: collector/nodes.go [hunk 9]
```diff
@@ -984,51 +525,14 @@ func (c *Nodes) Describe(ch chan<- *prometheus.Desc) {
 	ch <- c.jsonParseFailures.Desc()
 }
 
-func (c *Nodes) fetchAndDecodeNodeStats() (nodeStatsResponse, error) {
```
Comment (dominikschulz) [inline collector/nodes.go]: this method should remain in this file

File: collector/nodes_test.go [hunk 10]
```diff
@@ -1,102 +0,0 @@
...
-	"net/http"
-	"net/http/httptest"
-	"net/url"
-	"strings"
-	"testing"
-
-	"github.com/go-kit/kit/log"
-)
-
-func TestNodesStats(t *testing.T) {
```
Comment (dominikschulz) [inline collector/nodes_test.go]: this method should remain in this file

File: collector/metrics_collector.go [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+			"msg", "failed to fetch and decode index stats",
+			"err", err,
+		)
+		return
+	}
+	c.indices.up.Set(1)
+
+	c.clusterHealth.Collect(ch, clusterHealthResponse)
+	c.nodes.Collect(ch, nodeStatsResponse)
+	c.indices.Collect(ch, clusterHealthResponse, nodeStatsResponse, indexStatsResponse)
```
Comment (zwopir) [inline collector/metrics_collector.go]: wrapping collectors (or your custom non-interface-fulfilling Collect() ) isn't a good idea. Doing so makes the collectors run sequentially, not in parallel. Also the json retrieval works parallel, if you exclude it from the collectors. 
In additionMaking collectors optional is then just a matter of `if`ing the `prometheus.Mustregister()`.

Could you please refer to you initial intent of this PR: get the docs metrics of the primary shards. To archieve this I would strongly suggest to
 - go back to the start (sorry)
 - implement the indices collector (fulfilling the collector interface) with just the very few metrics included you are interested in. We would really like to keep the overall index metrics we can get via `_nodes/stats` in the nodes collector. Please only export the structs marshaled from the `_all/_stats` that are not in `_nodes/stats`.
 - add this collector optionally with `prometheus.MustRegister`

**Comment by matsumana** [comment]
I fixed.

Would you please review again?

File: collector/indices.go [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+	Value  func(indexStats IndexStatsIndexResponse) float64
+	Labels func(clusterName string, indexName string) []string
+}
+
+type Indices struct {
+	logger        log.Logger
+	client        *http.Client
+	url           *url.URL
+	all           bool
+	exportIndices bool
```
Comment (zwopir) [inline collector/indices.go]: please move the bool controlling if indices metrics should be scraped out of the collector. There is no reason registering the metrics via `Describe()` and then don't collect them. See my other comment in main.go

File: collector/indices.go [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+	}
+	ch <- i.up.Desc()
+	ch <- i.totalScrapes.Desc()
+	ch <- i.jsonParseFailures.Desc()
+}
+
+func (c *Indices) fetchAndDecodeIndexStats() (indexStatsResponse, error) {
+	var isr indexStatsResponse
+
+	if c.exportIndices {
```
Comment (zwopir) [inline collector/indices.go]: move if statement to `main.go`

File: main.go [hunk 14]
```diff
@@ -55,6 +56,7 @@ func main() {
 
 	prometheus.MustRegister(collector.NewClusterHealth(logger, httpClient, esURL))
 	prometheus.MustRegister(collector.NewNodes(logger, httpClient, esURL, *esAllNodes))
+	prometheus.MustRegister(collector.NewIndices(logger, httpClient, esURL, *esAllNodes, *esExportIndices))
```
Comment (zwopir) [inline main.go]: please replace by 
```
if *exExportIndices {
    prometheus.MustRegister(collector.NewIndices(logger, httpClient, esURL, *esAllNodes))
}
```
(so remove the flag from the NewIndices constructor as well)

File: collector/indices.go [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+	i.up.Set(1)
+
+	// Index stats
+	for indexName, indexStats := range indexStatsResponse.Indices {
+		for _, metric := range i.indexMetrics {
+			ch <- prometheus.MustNewConstMetric(
+				metric.Desc,
+				metric.Type,
+				metric.Value(indexStats),
+				metric.Labels(clusterHealthResponse.ClusterName, indexName)...,
```
Comment (zwopir) [inline collector/indices.go]: copy'n'paste error

Comment (zwopir) [reply collector/indices.go]: please review the labels you want to attach to the indices metrics. It seems there is not cluster name available

**Comment by matsumana** [comment]
@zwopir Thank you for the comments.
I fixed.
Would you please review again?

**Comment by zwopir** [comment]
looks good to me, thanks for your contribution!

**Pull request merged**